File: Question-2/q2-resource-management.kt
fun main() {
    val resourceManager = ResourceManager()
    resourceManager.addPatient(Patient(1, "Wanjiru", 2))
    resourceManager.addPatient(Patient(2, "Abrahim", 5))
    resourceManager.addPatient(Patient(3, "Wamai", 1))
    resourceManager.addStaff(Staff(1, "Dr. Sharon", "General"))
    resourceManager.addStaff(Staff(2, "Nurse Kimani", "Pediatrics"))
    resourceManager.addEquipment(Equipment(1, "X-Ray Machine", true))
    resourceManager.addEquipment(Equipment(2, "ECG machines", true))


    resourceManager.allocateResources()
}
data class Patient(val id: Int, val name: String, val urgency: Int)
data class Staff(val id: Int, val name: String, val expertise: String)
data class Equipment(val id: Int, val name: String, var available: Boolean)


class ResourceManager {
    private val patients = mutableListOf<Patient>()
    private val staff = mutableListOf<Staff>()
    private val equipment = mutableListOf<Equipment>()

    fun addPatient(patient: Patient) {
        patients.add(patient)
        println("Added patient: ${patient.name} with urgency level ${patient.urgency}")
    }

    fun addStaff(staffMember: Staff) {
        staff.add(staffMember)
        println("Added staff: ${staffMember.name} with expertise in ${staffMember.expertise}")
    }

    fun addEquipment(equip: Equipment) {
        equipment.add(equip)
        println("Added equipment: ${equip.name}")
    }

    fun allocateResources() {
        val sortedPatients = patients.sortedByDescending { it.urgency }

        for (patient in sortedPatients) {
            val availableStaff = staff.firstOrNull { it.expertise == "General" }
            val availableEquip = equipment.firstOrNull { it.available }

            if (availableStaff != null && availableEquip != null) {
                println("Allocating resources for patient: ${patient.name}")
                availableEquip.available = false 
            } else {
                println("No available resources for patient: ${patient.name}")
            }
        }
    }
}

File: Question-2/q2-resource-management.js
class PriorityQueue{
    constructor(){
 this.items= [];
    }
    enqueue(patient, priority){
        const element = {patient, priority};
        let added = false;
        for(let i = 0; i < this.items.length; i++){
            if (priority > this.items[i].priority){
                this.items.splice(i, 0, element);
                break;
            }
        }

        if(!added)this.items.push(element)
    }

    dequeue(){
        return this.items.shift()
    }

    isEmpty(){
        return this.items.length === 0;
    }

    peek(){
        return this.items[0];
    }
   

}

class Patient{
    constructor(id, urgencyLevel, arrivalTime, needs, requiredEquipment){
        this.id = id;
        this.urgencyLevel = urgencyLevel;
        this.arrivalTime = arrivalTime;
        this.needs = needs;
        this.requiredEquipment = requiredEquipment;
        this.assigned = false;
    }
}

class Bed{
    constructor(id, type, available= true){
        this.id = id;
        this.type = type;
        this.available = available;
    }
}

class Staff{
    constructor(id, name, specialization, available= true){
        this.id = id;
        this.name= name;
        this.specialization = specialization;
        this.available = available;
    }
}

class Equipment{
    constructor(id, type, available=true){
        this.id = id;
        this.type = type;
        this.available = available;
    }
}

class ResourceManager{
    constructor(){
    this.patients = [];
    this.beds = [];
    this.staff = [];
    this.equipment = [];
    this.waitingQueue = new PriorityQueue()
    }
    addPatient(patient) {
    this.patients.push(patient);
    this.waitingQueue.enqueue(patient, this.scorePatient(patient));

}

addBeds(bed){
    this.beds.push(bed)
}

addStaff(staff){
    this.staff.push(staff)

}

addEquipment(equipment){
    this.equipment.push(equipment);
}

scorePatient(patient){
    const now = Date.now()
    const waitingTime = (now - patient.arrivalTime)/1000
    const urgencyScore = 6 - patient.urgencyLevel;
    return urgencyScore * 1000 + waitingTime
    
}

allocateResources(){
    const allocations = []
    while(!this.waitingQueue.isEmpty()){

        const {patient} = this.waitingQueue.dequeue();
        if(patient.assigned)continue;
        const bed = this.findAvailableBed(patient);
        const staff = this.findAvailableStaff(patient);
        const equipment = this.findAvailableEquipment(patient);
        
if(bed && staff && equipment){
bed.available =false;
staff.available = false;
equipment.forEach(eq => (eq.available= false));
patient.assigned = true;
allocations.push({
    patientId: patient.id,
    bedId: bed.id,
    staffId: this.staff.id,
    equipmentIds: equipment.map(eq => eq.id)

});

}else{
    this.waitingQueue.enqueue(patient, this.scorePatient(this.patient));
    break;
}
}


return allocations;
}

findAvailableBed(patient){
    return this.beds.find(bed => bed.available && (bed.type === 'ICU' || bed.type === 'General'));
}

findAvailableStaff(patient){
    return this.staff.find(
        staff => staff.available && staff.specialization === patient.needs.specialization
    );
}

findAvailableEquipment(patient){
    if (!patient.requiredEquipment || patient.requiredEquipment.length === 0)return [];
    const found = [];
    for (const equipType of patient.requiredEquipment){
        const eq= this.equipment.find(e => e.available && e.type === equipType)
        if(!eq) return null;
        found.push(eq)
    }
    return found
}
handleSpikes(){
    this.beds.forEach(bed => (bed.available = true));
    this.staff.forEach(staff => (staff.available = true));
    this.equipment.forEach(eq => (eq.available = true));
}

}

const waiyakiHospital = new ResourceManager();
waiyakiHospital.addBeds(new Bed(1, "ICU"));
waiyakiHospital.addBeds(new Bed(2, "General"));
waiyakiHospital.addStaff(new Staff(1, "Dr. Pauline", "cardiology"));
waiyakiHospital.addStaff(new Staff(2, "Nurse Mutai", "general"));
waiyakiHospital.addEquipment(new Equipment(1, "ECG"));
waiyakiHospital.addEquipment(new Equipment(2, "X-ray"));
waiyakiHospital.addPatient(new Patient(1, 2, Date.now() - 10000, { specialization: "cardiology" }, ["X-ray"]));
waiyakiHospital.addPatient(new Patient(2, 4, Date.now() - 50000, { specialization: "general" }, ["ECG"]));

const allocations=waiyakiHospital.allocateResources();
console.log(allocations);

waiyakiHospital.handleSpikes();








